objc/objc_runtime.go

```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2022 The Ebitengine Authors
+
+//go:build darwin
+// +build darwin
+
+// Package objc is a low-level pure Go objective-c runtime. This package is easy to use incorrectly, so it is best
+// to use a wrapper that provides the functionality you need in a safer way.
+package objc
+
+import (
+	"fmt"
+	"reflect"
+	"runtime"
+	"unsafe"
+
+	"github.com/ebitengine/purego"
+	"github.com/ebitengine/purego/internal/strings"
+)
+
+//TODO: support try/catch?
+//https://stackoverflow.com/questions/7062599/example-of-how-objective-cs-try-catch-implementation-is-executed-at-runtime
+
+var (
+	objc = purego.Dlopen("/usr/lib/libobjc.A.dylib", purego.RTLD_GLOBAL)
+
+	objc_msgSend           = purego.Dlsym(objc, "objc_msgSend")
+	objc_msgSendSuper2     = purego.Dlsym(objc, "objc_msgSendSuper2")
+	objc_getClass          = purego.Dlsym(objc, "objc_getClass")
+	objc_allocateClassPair = purego.Dlsym(objc, "objc_allocateClassPair")
+	objc_registerClassPair = purego.Dlsym(objc, "objc_registerClassPair")
+	sel_registerName       = purego.Dlsym(objc, "sel_registerName")
+	class_getSuperclass    = purego.Dlsym(objc, "class_getSuperclass")
+	class_addMethod        = purego.Dlsym(objc, "class_addMethod")
+	object_getClass        = purego.Dlsym(objc, "object_getClass")
+)
+
+type Id uintptr
+
+func (id Id) Class() Class {
+	ret, _, _ := purego.SyscallN(object_getClass, uintptr(id))
+	return Class(ret)
+}
+
+// Send is a convenience method for sending messages to objects.
+func (id Id) Send(sel SEL, args ...interface{}) Id {
+	var tmp = make([]uintptr, len(args)+2)
+	createArgs(tmp, id, sel, args...)
+	ret, _, _ := purego.SyscallN(objc_msgSend, tmp...)
+	return Id(ret)
+}
+
+// SendSuper is a convenience method for sending message to object's super
+func (id Id) SendSuper(sel SEL, args ...interface{}) Id {
+	type objc_super struct {
+		reciever   Id
+		superClass Class
+	}
+	var _super = &objc_super{
+		reciever:   id,
+		superClass: id.Class(),
+	}
+	var tmp = make([]uintptr, len(args)+2)
+	createArgs(tmp, Id(unsafe.Pointer(_super)), sel, args...)
+	ret, _, _ := purego.SyscallN(objc_msgSendSuper2, tmp...)
+	return Id(ret)
+}
+
+func createArgs(out []uintptr, cls Id, sel SEL, args ...interface{}) {
+	if len(out) < len(args)+2 {
```

Comment:
What about making a slice inside `createArgs`?

```
out := make([]uintptr, len(args) + 2)
```

By the way, why is `len(out) < len(args)+2` required?

Reply:
The +2 is there because self and SEL are the first two arguments yet aren’t in the args slice

Reply:
Oh your asking about why check the length when it’s an internal method. It’s never too crazy to check input but I guess it’s not needed bc it’s internal

Reply:
I mean why we don't create a new slice inside createArgs.